### PR TITLE
Add ServerSocket read/write through selector

### PR DESCRIPTION
**Pull request by wggihan**

### CODE REVIEW

File: stdlib/io/src/main/java/org/ballerinalang/stdlib/io/channels/base/readers/AsyncReader.java [hunk 1]
```diff
@@ -36,7 +36,7 @@
  */
 public class AsyncReader implements Reader {
 
-    private static final Logger log = LoggerFactory.getLogger(BlockingReader.class);
+    private static final Logger log = LoggerFactory.getLogger(AsyncReader.class);
```
Comment (arukshani) [inline stdlib/io/src/main/java/org/ballerinalang/stdlib/io/channels/base/readers/AsyncReader.java]: Why is this class called AsynReader? Whether the read blocks or not depends on the channel this is been called on right? If the channel is in blocking mode won't the read method will block?

File: stdlib/io/src/main/java/org/ballerinalang/stdlib/io/channels/base/Channel.java [hunk 2]
```diff
@@ -138,15 +139,37 @@ public Channel(ByteChannel channel, Reader reader, Writer writer, int size) thro
...
      * Specifies whether the channel has reached to it's end.
      *
      * @return true if the channel has reached to it's end
      */
+    @Override
     public boolean hasReachedEnd() {
         return hasReachedToEnd;
     }
 
+    public ByteChannel getByteChannel() {
```
Comment (VijithaEkanayake) [inline stdlib/io/src/main/java/org/ballerinalang/stdlib/io/channels/base/Channel.java]: Don't we need to document this??

**Comment by VijithaEkanayake** [comment]
Can't we alter names of these classes CloseByteChannelEvent.java, ReadBytesEvent.java, WriteBytesEvent.java?? These names sound more like method names?? WDYT??

File: stdlib/io/src/main/java/org/ballerinalang/stdlib/io/events/records/DelimitedRecordReadEvent.java [hunk 3]
```diff
@@ -60,8 +63,16 @@ public DelimitedRecordReadEvent(DelimitedRecordChannel channel, EventContext con
     public EventResult get() {
         AlphaCollectionResult result;
         try {
-            String[] content = channel.read();
-            result = new AlphaCollectionResult(content, context);
+            if (channel.hasReachedEnd()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Channel " + channel.hashCode() + " reached it's end");
```
Comment (VijithaEkanayake) [inline stdlib/io/src/main/java/org/ballerinalang/stdlib/io/events/records/DelimitedRecordReadEvent.java]: We need to change it's to its' I think, Please check the all other relevant places.

File: tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/io/socket/ServerSocketTest.java [hunk 4]
```diff
@@ -53,59 +56,65 @@ public void setup() {
...
+            boolean connected = false;
+            for (int retryCount = 0; retryCount < numberOfRetryAttempts; retryCount++) {
+                try {
+                    Socket s = new Socket("localhost", port);
+                    BufferedReader input = new BufferedReader(new InputStreamReader(s.getInputStream()));
+                    String answer = input.readLine();
+                    Assert.assertEquals(welcomeMsg.trim(), answer, "Didn't get the expected response from server.");
+                    OutputStreamWriter out = new OutputStreamWriter(s.getOutputStream(), "UTF-8");
+                    out.write(clientMsg, 0, clientMsg.length());
+                    out.flush();
```
Comment (VijithaEkanayake) [inline tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/io/socket/ServerSocketTest.java]: Shouldn't we clean up the resource in a finally block?

**Pull request merged**